packages/flutter_tools/lib/src/mdns_discovery.dart

```diff
@@ -39,135 +44,385 @@ class MDnsObservatoryDiscovery {
   static MDnsObservatoryDiscovery? get instance => context.get<MDnsObservatoryDiscovery>();
 
   /// Executes an mDNS query for a Dart Observatory.
+  /// Checks for observatories that have already been launched.
+  /// If none found, will listen for new observatories to become active
+  /// and return first it finds that match parameters.
   ///
   /// The [applicationId] parameter may be used to specify which application
   /// to find. For Android, it refers to the package name; on iOS, it refers to
   /// the bundle ID.
   ///
-  /// If it is not null, this method will find the port and authentication code
+  /// The [deviceVmservicePort] parameter may be used to specify which port
+  /// to find.
+  ///
+  /// The [isNetworkDevice] parameter flags whether to get the device IP
+  /// and the [ipv6] parameter flags whether to get an iPv6 address
+  /// (otherwise it will get iPv4).
+  ///
+  /// The [timeout] parameter determines how long to continue to wait for
+  /// observatories to become active.
+  ///
+  /// If [applicationId] is not null, this method will find the port and authentication code
   /// of the Dart Observatory for that application. If it cannot find a Dart
-  /// Observatory matching that application identifier, it will call
+  /// Observatory matching that application identifier after the [timeout], it will call
   /// [throwToolExit].
   ///
-  /// If it is null and there are multiple ports available, the user will be
+  /// If [applicationId] is null and there are multiple ports available, the user will be
   /// prompted with a list of available observatory ports and asked to select
   /// one.
   ///
-  /// If it is null and there is only one available instance of Observatory,
-  /// it will return that instance's information regardless of what application
-  /// the Observatory instance is for.
+  /// If it is null and there is only one available or it's the first found instance
+  /// of Observatory, it will return that instance's information regardless of
+  /// what application the Observatory instance is for.
+  @visibleForTesting
+  Future<MDnsObservatoryDiscoveryResult?> queryForAttach({
+    String? applicationId,
+    int? deviceVmservicePort,
+    bool ipv6 = false,
+    bool isNetworkDevice = false,
+    Duration timeout = const Duration(minutes: 10),
+  }) async {
+    // Poll for 5 seconds to see if there are already services running.
+    // Use a new instance of MDnsClient so result don't get cached in _client.
```

Comment:
Why don't we want to cache the results?  Will the presence of anything in the cache cause it to not to try and lookup anything again?

Reply:
Exactly, so if it finds records in the first 5 second poll but it doesn't match the parameters (app id and/or device port), those results are cached and then in the next poll will end immediately because it gets from the cache.

https://github.com/flutter/packages/blob/62e04e1267c7f10cc00e6a4334213656b46d3964/packages/multicast_dns/lib/multicast_dns.dart#L208-L216